### PR TITLE
fine tune preserve logins

### PR DESCRIPTION
**Pull request by brindy**
Task/Issue URL: https://app.asana.com/0/392891325557410/1164380708432526
Tech Design URL: https://app.asana.com/0/392891325557410/1164681539353129
CC:

**Description**:

Adds setting to detect logins and if on, prompt user to fireproof website and includes refactor to pull user script and message handling in to their own classes.

General expectations are
* that false positives (ie detecting a login when there isn't one) is such that there should be very few (or ideally none).  
* false positives while failing to attempt a genuine login are acceptable.
* false negatives (not detecting a login when there is one) will have increased, but this is OK as the user can still use the Fireproof option in the menu.
* if a site does not use cookies exclusively then fireproofing may not work

This functionality is only available for iOS 13.

**Steps to test this PR**:

***Regression testing:***

1. Ensure tracker blocking still works as expected
1. Ensure when debugging that debugging scripts still work as expected
1. Ensure DocumentUserScript only injected on iOS < 13
1. Ensure FindInPage still works

***Functionality testing:***

1. Enable login detection (Settings > Fireproof Websites > Toggle)
1. Start logging in to sites - user should be prompted to fireproof
1. Check that selecting to Fireproof adds site to the list and that the user remains logged in after using the fire button

1. Disable login detection and remove any fireproofed sites.
1. Log in to the sites that were detected previously.
1. Check the user is not prompted to Fireproof the site.

1. Ensure when changing toggle state that tabs are reloaded with correct user script configuration.

1. Ensure setting only available on iOS 13.

1. Ensure pixel fired when changing login detection state toggle.

**Device Testing**:

* [x] iPhone SE
* [x] iPhone 8
* [x] iPhone X
* [x] iPad

**OS Testing**:

* [x] iOS 10
* [x] iOS 11
* [x] iOS 12
* [x] iOS 13

---
###### Internal references:
[Software Engineering Expectations](https://app.asana.com/0/59792373528535/199064865822552)
[Technical Design Template](https://app.asana.com/0/59792373528535/184709971311943)

### CODE REVIEW

File: Core/document.js [hunk 1]
```diff
@@ -25,19 +25,16 @@ var duckduckgoDocument = function () {
...
         }
-        return null
+        
+        return null;
     };
     
     return {
-        getHrefFromPoint: getHrefFromPoint,
-        getHrefFromElement: getHrefFromElement
+        getHrefFromPoint: getHrefFromPoint
```
Comment (brindy) [inline Core/document.js:37]: Removed some unused code and optimised slightly.

File: DuckDuckGo/BrokenSiteInfo.swift [hunk 2]
```diff
@@ -73,7 +73,7 @@ public struct BrokenSiteInfo {
         guard let url = url else { return "" }
         
         var components = URLComponents(url: url, resolvingAgainstBaseURL: true)
-        components?.queryItems = []
+        components?.queryItems = nil
```
Comment (brindy) [inline DuckDuckGo/BrokenSiteInfo.swift:76]: Spotted a bug with trailing ? in broken site monitoring - minor fix

File: DuckDuckGo/TabViewController.swift [hunk 3]
```diff
@@ -820,6 +727,9 @@ extension TabViewController: WKNavigationDelegate {
     
     private func onWebpageDidStartLoading(httpsForced: Bool) {
         os_log("webpageLoading started", log: generalLog, type: .debug)
+        
+        // isLoginFormDetected = false
+
```
Comment (bwaresiak) [inline DuckDuckGo/TabViewController.swift]: Is this needed? :)

**Review by bwaresiak — COMMENTED** [review]
I'm still in the process of testing, few comments:
1. I'd consider adding detail label (Auto/Manual or something like that) to Fireproof settings cell, as there is no indicator there is a toggle underneath there. Shall it be enabled or disabled by default?
2. I'm worried people may misunderstand how fireproofing works and have an expectation that removing given website from fireproof list will actually log out them from that site immediately. Have we considered that?

File: .github/PULL_REQUEST_TEMPLATE.md [hunk 4]
```diff
@@ -13,6 +13,20 @@ CC:
...
+* [ ] iPhone X
+* [ ] iPad
+
+**OS Testing**:
+
+* [ ] iOS 10
+* [ ] iOS 11
+* [ ] iOS 12
+* [ ] iOS 13
+
```
Comment (bwaresiak) [inline .github/PULL_REQUEST_TEMPLATE.md:29]: Great, maybe it would be good to add Dark/Light themes as well?

**Review by bwaresiak — APPROVED** [review]
Great work :)

**Comment by brindy** [comment]
Thanks @bwaresiak !

**Pull request merged**